third_party/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -682,6 +686,33 @@ where
         }
     }
 
+    // Check whether the values on the operand stack have the expected return types.
+    // The given types will also be replaced by the expected type, thus if a given
+    // type is more specific (assignable) then expected, it will be replaced by the
+    // type known for the value in the caller context.
+    fn check_and_replace_return_tys(&mut self, current_frame: &mut Frame) -> PartialVMResult<()> {
+        let expected_ret_tys = current_frame.function.return_tys();
+        if expected_ret_tys.is_empty() {
+            return Ok(());
+        }
+        let given_ret_tys = self.operand_stack.last_n_tys_mut(expected_ret_tys.len())?;
+        for (expected, given) in expected_ret_tys.iter().zip(given_ret_tys) {
+            let ty_args = current_frame.function.ty_args();
+            if ty_args.is_empty() {
+                given.paranoid_check_assignable(expected)?;
+                *given = expected.clone()
+            } else {
+                let expected_inst = self
```

Comment:
This is duplicated with function call check for param tys. I am not convinced we need this replacement if all callers check assignability anyway?

Reply:
We've discussed that we don't feel comfortable with current treatment, though it (might) still be ok. Consider:

```
fun f(): |u64|u32 { action /* has copy+drop */ }
fun g(x: |u64|u32 has copy+drop)
...
g(f())
```

This passes runtime checks, so caller does _not_ check the constraint of `f` return type. It would if there is a local for the return value of `f`, but this doesn't have to be like this. While it might not be harmful (at least as far as we know, and as of today), it's weird.

Reply:
It is type checked. However, as I showed in the above example, the actual return value of `f` might be `bool` and `g` is in fact expecting a `bool` so that check succeeds. Nevertheless, `f` might have been _declared_ to have type `u64`, which is simply ignored here. While we didn't found a way how this can be exploited it feels brittle and problematic for potential future changes. We don't want to allow `f` to return something else as declared.